smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java

```diff
@@ -1,13 +1,150 @@
 package com.eis.smslibrary;
 
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
 import com.eis.communication.CommunicationHandler;
-import com.eis.communication.Message;
+import com.eis.smslibrary.listeners.SMSReceivedListener;
+import com.eis.smslibrary.listeners.SMSSentListener;
+
+import java.lang.ref.WeakReference;
+
+/**
+ * Communication handler for SMSs. It's a Singleton, you should
+ * access it with {@link #getInstance}, and before doing anything you
+ * should call {@link #setup}.
+ *
+ * @author Luca Crema, Marco Mariotto, Alberto Ursino, Marco Tommasini
+ */
+public class SMSHandler implements CommunicationHandler<SMSMessage> {
+
+    public static final String SENT_MESSAGE_INTENT_ACTION = "SMS_SENT";
+    public static final int RANDOM_STARTING_COUNTER_VALUE_RANGE  = 100000;
+
+    /**
+     * Singleton instance
+     */
+    private static SMSHandler instance;
+
+    /**
+     * Weak reference doesn't prevent garbage collector to
+     * de-allocate this class when it has reference to a
+     * context that is still running. Prevents memory leaks.
+     */
+    private WeakReference<Context> context;
 
+    /**
+     * Received listener reference
+     */
+    private SMSReceivedListener receivedListener;
+
+    /**
+     * This message counter is used so that we can have a different action name
+     * for pending intent (that will call broadcastReceiver). If we were to use the
+     * same action name for every message we would have a conflict and we wouldn't
+     * know what message has been sent
+     */
+    private int messageCounter;
+
+    /**
+     * Private constructor for Singleton
+     */
+    private SMSHandler() {
+        //Random because if we close and open the app the value probably differs
+        messageCounter = (int)(Math.random() * RANDOM_STARTING_COUNTER_VALUE_RANGE);
+    }
+
+    /**
+     * @return the current instance of this class
+     */
+    public static SMSHandler getInstance() {
+        if (instance == null)
+            instance = new SMSHandler();
+
+        return instance;
+    }
 
-public class SMSHandler implements CommunicationHandler {
+    /**
+     * Setup for the handler.
+     *
+     * @param context current context.
+     */
+    public void setup(Context context) {
+        this.context = new WeakReference<>(context);
+    }
 
+    /**
+     * Sends a message to a destination peer via SMS.
+     * Requires {@link android.Manifest.permission#SEND_SMS}
+     *
+     * @param message to be sent in the channel to a peer
+     */
     @Override
-    public void sendMessage(Message message) {
-        //Test pull request
+    public void sendMessage(final @NonNull SMSMessage message) {
+        sendMessage(message, null);
```

Comment:
Why is the checkSetup not here?

Reply:
Because it is done in the other method, this is just an overload

Reply:
Ah ok I thought this method was done first...

Reply:
@CremaLuca

Reply:
Anyway I think it's a very small and almost useless addition / change so you see what to do.

Reply:
It would be a repetition, the check is already made in the `sendMessage(SMSMessage, Listener)`, adding the `checkSetup()` again would be just a redundant check, because the excecution order would become
- call checkSetup
- call sendMessage(..., ...)
- call checkSetup (again)